src/main/java/seedu/java/util/TaskList.java

```diff
@@ -0,0 +1,134 @@
+    private String path = "data/duke.txt";
+
+    public TaskList() {
+        this.taskArr = new ArrayList<>();
+    }
+
+    public TaskList(ArrayList<Task> taskArr) {
+        this.taskArr = taskArr;
+    }
+
+    /**
+     * Returns its ArrayList of Tasks
+     * @return taskArr
+     */
+    public ArrayList<Task> getTaskArr() {
+        return taskArr;
+    }
+
+    private String totalTask() {
+        return "You have " + (taskArr.size()) + " task on your list.";
+    }
+
+    private String addTodo(String task) {
+        taskArr.add(new Todo(task));
+        return "Affirmative. Adding a to-do task :) \n" + "added: " + task + "\n" + totalTask();
+    }
+
+    private String addDead(String task, String timing) {
+        taskArr.add(new Deadline(task, timing));
+        return "Affirmative. Adding a task with Deadline :) \n" + "added: " + task + "\n" + totalTask();
+    }
+
+    private String addEvent(String task, String timing) {
+        taskArr.add(new Event(task, timing));
+        return "Affirmative. Adding an event :) \n" + "added: " + task + "\n" + totalTask();
+    }
+
+    private String deleteTask(int index) {
+        taskArr.remove(index - 1);
+        return "Deleting this task. now you have less things to do :)" + "\n" + totalTask();
+    }
+
+    /**
+     * Completes a task and returns a string to be printed by the UI.
+     * @param index - the task that you want printed
+     * @return String to print
+     */
+    private String doneTask(int index) {
+        taskArr.set(index - 1, taskArr.get(index - 1).completeTask());
+        return "Complete task! :)" + "\n" + totalTask();
+    }
+
+    /**
+     * Returns a String of list of Task.
+     * @return String of list
+     */
+    public String listToPrint() {
+        int i = 1;
+        String toPrint = "";
+        for (Task x : taskArr) {
+            toPrint += (i++) + "" + x.toString() + "\n";
+        }
+        return toPrint;
+    }
+
+    private String listToPrint(ArrayList<Task> tasks) {
+        int i = 1;
+        String toPrint = "";
+        for (Task x : tasks) {
+            toPrint += (i++) + "" + x.toString() + "\n";
+        }
+        return toPrint;
+    }
+
+    private String findTask(String keyword) {
+        ArrayList<Task> matchedTask = new ArrayList<>();
+        for (Task x: taskArr) {
+            if(x.getTask().contains(keyword)) {
+                matchedTask.add(x);
+            }
+        }
+        if (matchedTask.isEmpty()) {
+            return "I'm sorry, but your search garners no results";
+        } else {
+            return listToPrint(matchedTask);
+        }
+    }
+
+    /**
+     * Reads an input from UI, performs an operation, and returns a string to be printed.
+     * @param input from user
+     * @return String to print for confirmation
+     */
+    public String read(String input) {
+        try {
+            Command cmd = Parser.readCommand(input);
+            switch (cmd) {
```

Comment:
To me, this switch statement feels like it could be improved, as it is another potential violation of Open-Closed Principle. Adding a new command will require a new `case`, and even a new method in class `TaskList`!

It would be quite tedious to solve this issue though. One way I have in mind would be to expand `Command` into a full class, and have it implement a method `executeCommand(ArrayList<Task> tasksList, String task, String timing, ...)` (you can use `String[]` or `String...` for the many Strings) so you would just call `executeCommand` here, regardless of what the command actually is. But that would require an overhaul of a large portion of the code

Any thoughts?

Reply:
This is a great comment! Now it makes more sense as to why one of the increments suggested adding Commands and its subclasses now. At that point I didn't see much sense in it, but now, maybe I'll integrate that into my own Duke!